File: src/python/problems/remove_element_test.py
# https://leetcode.com/problems/remove-element

from core.problem_base import *

class RemoveElement(ProblemBase):
    def Solution(self, nums: list[int], val: int) -> int:
        left = 0
        for right in range(len(nums)):
            if nums[right] != val:
                nums[left] = nums[right]
                left += 1 

        return (left, nums[:left])

if __name__ == '__main__':
    TestGen(RemoveElement) \
        .Add(lambda tc: tc.Param([0,1,2,2,3,0,4,2]).Param(2).Result((5, [0,1,3,0,4]))) \
        .Add(lambda tc: tc.Param([3,2,2,3]).Param(3).Result((2, [2,2]))) \
        .Run()

File: src/go/problems/remove_element_test.go
// https://leetcode.com/problems/remove-element

package problems

import (
	"testing"
	"github.com/dobermanch/leetcode/core"
)

type RemoveElement struct{}

func TestRemoveElement(t *testing.T) {
	gen := core.TestSuite[RemoveElement]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([]int{3,2,2,3}).Param(3).Result(2)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{0,1,2,2,3,0,4,2}).Param(2).Result(5)
	}).Run(t)
}

func (RemoveElement) Solution(nums []int, val int) int {
    left := 0
    for right := 0; right < len(nums); right++ {
        if nums[right] != val {
            nums[left] = nums[right]
            left++
        }
    }

    return left
}